### PR TITLE
A `CURRENT_{STRUCT,VARIANT}` wrapper over `rapidjson`.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: blocks/json/json.h [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  bool empty() const { return keys.empty(); }
+  bool HasKey(const std::string& key) const { return fields.count(key); }
+  const JSONValue& operator[](const std::string& key) const {
+    static JSONValue null((JSONNull()));
+    const auto cit = fields.find(key);
+    return cit != fields.end() ? cit->second : null;
+  }
+  JSONObject& push_back(const std::string& name, const JSONValue& value) {
+    keys.push_back(name);
+    fields.emplace(name, value);
```
Comment (mzhurovich) [inline blocks/json/json.h]: @dkorolev Hmm, what if the key already exists?

File: blocks/json/test.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  EXPECT_EQ(json, AsJSON(mutable_parsed));
+  Value<JSONObject>(mutable_parsed).erase("z");
+  EXPECT_EQ(json, AsJSON(mutable_parsed));
+  Value<JSONObject>(mutable_parsed).erase("a");
+  EXPECT_EQ("{\"c\":\"the order is preserved\",\"b\":null}", AsJSON(mutable_parsed));
+  Value<JSONObject>(mutable_parsed).push_back("a", JSONNumber(101));
+  EXPECT_EQ("{\"c\":\"the order is preserved\",\"b\":null,\"a\":101}", AsJSON(mutable_parsed));
+  Value<JSONObject>(mutable_parsed).erase("b").erase("c").push_back("c", JSONString("foo")).push_back("b", JSONNull());
+  EXPECT_EQ("{\"a\":101,\"c\":\"foo\",\"b\":null}", AsJSON(mutable_parsed));
+
```
Comment (mzhurovich) [inline blocks/json/test.cc:115]: Test for adding the same key twice?

**Comment by dkorolev** [comment]
@mzhurovich good point, fixed in 6e9658ef.

File: blocks/json/json.h [hunk 3]
```diff
@@ -162,17 +162,13 @@ CURRENT_STRUCT(JSONObject) {
...
-      auto it_list_next = it_list;
-      while (++it_list_next != std::end(keys)) {
-        *it_list = *it_list_next;
-        ++it_list;
-      }
-      *it_list = name;
       it_field->second = value;
+      auto it_list = std::find(std::begin(keys), std::end(keys), name);
+      std::rotate(it_list, it_list + 1u, std::end(keys));
+      keys.back() = name;
```
Comment (mzhurovich) [inline blocks/json/json.h]: This line seems redundant. Feel free to merge after fixing :)

Comment (dkorolev) [reply blocks/json/json.h]: :+1:

**Pull request merged**